identity-did/src/verification/verification_method.rs

```diff
@@ -104,37 +152,139 @@ impl<T> VerificationMethod<T> {
     &mut self.properties
   }
 
-  pub fn try_into_fragment(&self) -> Result<String> {
-    self
-      .id
-      .fragment()
-      .ok_or(Error::InvalidMethodFragment)
-      .map(|fragment| once('#').chain(fragment.chars()).collect())
-  }
-
   /// Creates a new [`MethodRef`] from `self`.
-  pub fn into_ref(self) -> MethodRef<T> {
+  pub fn into_method_ref(self) -> MethodRef<D, T> {
     MethodRef::Embed(self)
   }
+
+  /// Maps `VerificationMethod<D,T>` to `VerificationMethod<C,T>` by applying a function `f` to
+  /// the id and controller.
+  pub fn map<C, F>(self, mut f: F) -> VerificationMethod<C, T>
+  where
+    C: DID,
+    F: FnMut(D) -> C,
+  {
+    VerificationMethod {
+      id: self.id.map(&mut f),
+      controller: f(self.controller),
+      key_type: self.key_type,
+      key_data: self.key_data,
+      properties: self.properties,
+    }
+  }
+
+  /// Fallible version of [`VerificationMethod::map`].
+  pub fn try_map<C, F, E>(self, mut f: F) -> Result<VerificationMethod<C, T>, E>
+  where
+    C: DID,
+    F: FnMut(D) -> Result<C, E>,
+  {
+    Ok(VerificationMethod {
+      id: self.id.try_map(&mut f)?,
+      controller: f(self.controller)?,
+      key_type: self.key_type,
+      key_data: self.key_data,
+      properties: self.properties,
+    })
+  }
 }
 
-impl<T> Display for VerificationMethod<T>
+impl<D, T> VerificationMethod<D, T>
 where
+  D: DID,
+  T: Revocation + Default,
+{
+  // ===========================================================================
+  // Constructors
+  // ===========================================================================
+
+  /// Creates a new [`IotaVerificationMethod`] from the given `did` and public key.
+  pub fn new(did: D, key_type: KeyType, public_key: &PublicKey, fragment: &str) -> Result<Self> {
+    let method_fragment: String = if !fragment.starts_with('#') {
+      format!("#{}", fragment)
+    } else {
+      fragment.to_owned()
+    };
+    let id: DIDUrl<D> = did.to_url().join(method_fragment)?;
```

Comment:
I really dislike `Fragment` since it does not enforce character restrictions on the string contents, so you can put spaces or whatever invalid unicode you like into it without validation.

I'll add `Fragment` to my list of things to remove or refactor but I wouldn't use it here, even if `DIDUrl` would validate the contents correctly.

Reply:
That's reasonable. I agree it's better to remove `Fragment` entirely then.